File: src/main/scala/s/strategy/Player.scala
/* (C) 2022 */
package s.strategy

class Player(var name: String, var strategy: Strategy) {
  private var wincount = 0
  private var losecount = 0
  private var gamecount = 0

  def nextHand: Hand = {
    val result = strategy.nextHand
    strategy = result._1
    result._2
  }

  def win(): Unit = {
    strategy = strategy.study(true)
    wincount += 1
    gamecount += 1
  }

  def lose(): Unit = {
    strategy = strategy.study(false)
    losecount += 1
    gamecount += 1
  }

  def even(): Unit = {
    gamecount += 1
  }

  override def toString: String =
    "[" + name + ":" + gamecount + " games, " + wincount + " win, " + losecount + " lose" + "]"
}

File: src/main/kotlin/k/strategy/Player.kt
package k.strategy

class Player(private val name: String, private var strategy: Strategy) {
  private var wincount = 0
  private var losecount = 0
  private var gamecount = 0

  fun nextHand(): Hand {
    val result = strategy.nextHand()
    strategy = result.first
    return result.second
  }

  fun win() {
    strategy = strategy.study(true)
    wincount++
    gamecount++
  }

  fun lose() {
    strategy = strategy.study(false)
    losecount++
    gamecount++
  }

  fun even() {
    gamecount++
  }

  override fun toString(): String {
    return ("[" +
        name +
        ":" +
        gamecount +
        " games, " +
        wincount +
        " win, " +
        losecount +
        " lose" +
        "]")
  }
}
